pkg/controller/serviceaccount/legacy_serviceaccount_token_cleaner.go

```diff
@@ -0,0 +1,274 @@
+	"k8s.io/kubernetes/pkg/serviceaccount"
+)
+
+const (
+	dateFormat = "2006-01-02"
+)
+
+// TokenCleanerOptions contains options for the TokenCleaner
+type LegacySATokenCleanerOptions struct {
+	// LegacySATokenCleanUpInterval is the time interval that the token cleaner re-evaluates whether a legacy token should be deleted.
+	LegacySATokenCleanUpInterval time.Duration
+
+	// LegacySATokenCleanUpPeriod is the period of time since the last usage of an legacy token before it can be deleted.
+	LegacySATokenCleanUpPeriod time.Duration
+}
+
+// LegacySATokenCleaner is a controller that deletes legacy serviceaccount tokens that are not in use for a specified period of time.
+type LegacySATokenCleaner struct {
+	client clientset.Interface
+
+	saLister         listersv1.ServiceAccountLister
+	saInformerSynced cache.InformerSynced
+
+	secretLister         listersv1.SecretLister
+	secretInformerSynced cache.InformerSynced
+
+	podLister         listersv1.PodLister
+	podInformerSynced cache.InformerSynced
+
+	configMapLister         listersv1.ConfigMapLister
+	configMapInformerSynced cache.InformerSynced
+
+	legacySATokenCleanUpInterval time.Duration
+	legacySATokenCleanUpPeriod   time.Duration
+}
+
+// NewLegacySATokenCleaner returns a new *NewLegacySATokenCleaner.
+func NewLegacySATokenCleaner(saInformer coreinformers.ServiceAccountInformer, secretInformer coreinformers.SecretInformer, podInformer coreinformers.PodInformer, configMapInformer coreinformers.ConfigMapInformer, cl clientset.Interface, options LegacySATokenCleanerOptions) (*LegacySATokenCleaner, error) {
+	tc := &LegacySATokenCleaner{
+		client:                       cl,
+		saLister:                     saInformer.Lister(),
+		saInformerSynced:             saInformer.Informer().HasSynced,
+		secretLister:                 secretInformer.Lister(),
+		secretInformerSynced:         secretInformer.Informer().HasSynced,
+		configMapLister:              configMapInformer.Lister(),
+		configMapInformerSynced:      configMapInformer.Informer().HasSynced,
+		podLister:                    podInformer.Lister(),
+		podInformerSynced:            podInformer.Informer().HasSynced,
+		legacySATokenCleanUpPeriod:   options.LegacySATokenCleanUpPeriod,
+		legacySATokenCleanUpInterval: options.LegacySATokenCleanUpInterval,
+	}
+
+	return tc, nil
+}
+
+func (tc *LegacySATokenCleaner) Run(ctx context.Context) {
+	defer utilruntime.HandleCrash()
+
+	logger := klog.FromContext(ctx)
+	logger.Info("Starting legacy service account token cleaner controller")
+	defer logger.Info("Shutting down legacy service account token cleaner controller")
+
+	if !cache.WaitForNamedCacheSync("legacy-service-account-token-cleaner", ctx.Done(), tc.saInformerSynced, tc.secretInformerSynced, tc.configMapInformerSynced, tc.podInformerSynced) {
+		return
+	}
+
+	go wait.UntilWithContext(ctx, tc.evaluateSATokens, tc.legacySATokenCleanUpInterval)
+
+	<-ctx.Done()
+}
+
+func (tc *LegacySATokenCleaner) evaluateSATokens(ctx context.Context) {
```

Comment:
the first step should be to check the configmap, and see if we've even been tracking use long enough to know if a secret is unused for long enough:

```go
now := tc.clock.Now().UTC()
trackedSince, err := ... get date from configmap (00:00 on day _after_ the date in `since` key)
if err != nil {
  return
}
if now < trackedSince.Add(tc.minimumSinceLastUsed) {
  // we haven't been tracking long enough
  return
}
```

the second step should be to compute the earliest last-used or creationTimestamp we should preserve
```
preserveCreatedOnOrAfter := now.Add(-tc.minimumSinceLastUsed)
preserveUsedOnOrAfter := now.Add(-tc.minimumSinceLastUsed).Format(...)
```

then iterate through the secrets and `continue` early if:
* `secret.Type != v1.SecretTypeServiceAccountToken`
* `secret.CreationTimestamp >= preserveCreatedOnOrAfter`
* `lastUsed, ok := secret.Labels["kubernetes.io/legacy-token-last-used"]; ok && lastUsed >= preserveUsedOnOrAfter`